File: go/test/info_ip_get_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/internetdb-sdk/go"
	"github.com/voxgig-sdk/internetdb-sdk/go/core"

	vs "github.com/voxgig-sdk/internetdb-sdk/go/utility/struct"
)

func TestInfoIpGetEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.InfoIpGet(nil)
		if ent == nil {
			t.Fatal("expected non-nil InfoIpGetEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := info_ip_getBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "info_ip_get." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set INTERNETDB_TEST_INFO_IP_GET_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		infoIpGetRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.info_ip_get", setup.data)))
		var infoIpGetRef01Data map[string]any
		if len(infoIpGetRef01DataRaw) > 0 {
			infoIpGetRef01Data = core.ToMapAny(infoIpGetRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = infoIpGetRef01Data

		// LIST
		infoIpGetRef01Ent := client.InfoIpGet(nil)
		infoIpGetRef01Match := map[string]any{
			"ip": setup.idmap["ip01"],
		}

		infoIpGetRef01ListResult, err := infoIpGetRef01Ent.List(infoIpGetRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		_, infoIpGetRef01ListOk := infoIpGetRef01ListResult.([]any)
		if !infoIpGetRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", infoIpGetRef01ListResult)
		}

	})
}

func info_ip_getBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "info_ip_get", "InfoIpGetTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read info_ip_get test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse info_ip_get test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"info_ip_get01", "info_ip_get02", "info_ip_get03", "ip01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("INTERNETDB_TEST_INFO_IP_GET_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"INTERNETDB_TEST_INFO_IP_GET_ENTID": idmap,
		"INTERNETDB_TEST_LIVE":      "FALSE",
		"INTERNETDB_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["INTERNETDB_TEST_INFO_IP_GET_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["INTERNETDB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewInternetdbSDK(core.ToMapAny(mergedOpts))
	}

	live := env["INTERNETDB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["INTERNETDB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/info_ip_get_entity_test.lua
-- InfoIpGet entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("internetdb_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("InfoIpGetEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:InfoIpGet(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["info_ip_get"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:InfoIpGet(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:InfoIpGet(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = info_ip_get_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"list"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "info_ip_get." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set INTERNETDB_TEST_INFO_IP_GET_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local info_ip_get_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.info_ip_get")))
    local info_ip_get_ref01_data = nil
    if #info_ip_get_ref01_data_raw > 0 then
      info_ip_get_ref01_data = helpers.to_map(info_ip_get_ref01_data_raw[1][2])
    end

    -- LIST
    local info_ip_get_ref01_ent = client:InfoIpGet(nil)
    local info_ip_get_ref01_match = {
      ["ip"] = setup.idmap["ip01"],
    }

    local info_ip_get_ref01_list_result, err = info_ip_get_ref01_ent:list(info_ip_get_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(info_ip_get_ref01_list_result)

  end)
end)

function info_ip_get_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/info_ip_get/InfoIpGetTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read info_ip_get test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "info_ip_get01", "info_ip_get02", "info_ip_get03", "ip01" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("INTERNETDB_TEST_INFO_IP_GET_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["INTERNETDB_TEST_INFO_IP_GET_ENTID"] = idmap,
    ["INTERNETDB_TEST_LIVE"] = "FALSE",
    ["INTERNETDB_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["INTERNETDB_TEST_INFO_IP_GET_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["INTERNETDB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["INTERNETDB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["INTERNETDB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
